File: 백준/Gold/1113. 수영장 만들기/수영장 만들기.java
import java.io.*;
import java.util.*;

public class Main {

    static int n, m;
    static int[][] area;
    static boolean[][] visited;
    static int[] di = {0, 1, 0, -1};
    static int[] dj = {1, 0, -1, 0};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        area = new int[n][m];
        for (int i = 0; i < n; i++) {
            String line = br.readLine();
            for (int j = 0; j < m; j++) {
                area[i][j] = line.charAt(j) - '0';
            }
        }

        int ans = 0;
        for (int h = 2; h <= 9; h++) {
            visited = new boolean[n][m];
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (area[i][j] < h && !visited[i][j]) {
                        int temp = bfs(i, j, h);
                        if (temp != -1) {
                            ans += temp;
                        }
                    }
                }
            }
        }

        System.out.println(ans);
    }

    static int bfs(int i, int j, int h) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[] {i, j});
        visited[i][j] = true;

        List<int[]> conn = new ArrayList<>();
        conn.add(new int[] {i, j});

        boolean adjOutline = false;

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int x = cur[0];
            int y = cur[1];

            for (int d = 0; d < 4; d++) {
                int nx = x + di[d];
                int ny = y + dj[d];

                if (nx >= 0 && nx < n && ny >= 0 && ny < m) {
                    if (!visited[nx][ny] && area[nx][ny] < h) {
                        q.add(new int[] {nx, ny});
                        visited[nx][ny] = true;
                        conn.add(new int[] {nx, ny});
                    }
                } else {
                    adjOutline = true;
                }
            }
        }

        if (adjOutline) {
            return -1;
        }

        int res = 0;
        for (int[] p : conn) {
            int x = p[0], y = p[1];
            res += h - area[x][y];
            area[x][y] = h;
        }

        return res;
    }
}

File: 백준/Gold/1113. 수영장 만들기/수영장 만들기.py
import sys
from collections import deque

n, m = map(int, input().split())
area = [list(map(int, input())) for _ in range(n)]

di = [0, 1, 0, -1]
dj = [1, 0, -1, 0]

def bfs(i, j, h):
    q = deque()
    q.append([i, j])
    visited[i][j] = True

    conn = [[i, j]]
    adj_outline = False
    while q:
        x, y = q.popleft()

        for k in range(4):
            nx, ny = x + di[k], y + dj[k]
            if 0 <= nx < n and 0 <= ny < m :
                if not visited[nx][ny] and area[nx][ny] < h:
                    q.append([nx, ny])
                    visited[nx][ny] = True
                    conn.append([nx, ny])
            else:
                adj_outline = True

    if adj_outline:
        return -1

    res = 0
    for x, y in conn:
        res += h - area[x][y]
        area[x][y] = h

    return res


ans = 0
for h in range(2, 10):
    visited = [[False] * m for _ in range(n)]
    for i in range(n):
        for j in range(m):
            if area[i][j] < h and not visited[i][j]:
                # 붙어있는 칸중에 높이가 같은 애들을 다 구해
                temp = bfs(i, j, h)

                if temp == -1:
                    continue
                else:
                    ans += temp


print(ans)